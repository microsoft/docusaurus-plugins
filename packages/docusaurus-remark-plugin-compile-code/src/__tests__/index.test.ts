import { describe, it, expect } from 'vitest';
import path from 'path';
import vfile from 'to-vfile';
import mdx from 'remark-mdx';
import remark from 'remark';
import plugin from '../index';
import type { PluginOptions } from '../types';
import { readFileSync } from 'fs-extra';

const processFixture = async (name: string, options: PluginOptions) => {
  const filePath = path.join(__dirname, '__fixtures__', `${name}.md`);
  const file = await vfile.read(filePath);
  const result = await remark().use(mdx).use(plugin, options).process(file);

  return result.toString();
};

const options: PluginOptions = {
  langs: [
    {
      lang: 'ts',
      nodeBin: 'tsc',
      inputFiles: {
        'tsconfig.json': {
          compilerOptions: {
            target: 'esnext',
            skipLibCheck: true,
            lib: ['esnext', 'dom'],
          },
          include: ['input.ts'],
        },
      },
    },
    {
      lang: 'meta',
      meta: 'bar',
      compile: async (source, langOptions) => ({
        stdout: source,
      }),
    },
    {
      lang: 'fail',
      errorLang: 'console',
      errorMeta: 'console',
      meta: 'foo',
      compile: async (source, langOptions) => {
        throw new Error('fail');
      },
    },
    {
      lang: 'echo',
      prefix: 'prefix',
      inputLang: 'lisp',
      compile: async (source, langOptions) => ({
        stdout: source,
      }),
    },
    {
      lang: 'echonull',
      inputLang: null,
      outputLang: null,
      compile: async (source, langOptions) => ({
        stdout: source,
      }),
    },
    {
      lang: 'echonode',
      compile: async (source, langOptions) => ({
        nodes: [
          {
            type: 'code',
            meta: 'extra',
            value: 'extra --> ' + source,
          },
          {
            type: 'code',
            lang: 'json',
            value: JSON.stringify(langOptions, null, 2),
          },
        ],
      }),
    },
  ],
};

describe('compile-code plugin', () => {
  it('works on compile file', async () => {
    const result = await processFixture('compile', options);
    expect(result).toMatchSnapshot();
  });
  it('works on nodebin file', async () => {
    const result = await processFixture('nodebin', options);
    expect(result).toMatchSnapshot();
  });
  it('works on meta file', async () => {
    const result = await processFixture('meta', options);
    expect(result).toMatchSnapshot();
  });
  it('works on nodes file', async () => {
    const result = await processFixture('nodes', options);
    expect(result).toMatchSnapshot();
  });
  it('works on puppet file', async () => {
    const options: PluginOptions = {
      langs: [
        {
          lang: 'puppet',
          createDriverHtml: (options) => {
            const filePath = path.join(__dirname, '__fixtures__', `puppet.html`);
            return readFileSync(filePath, { encoding: 'utf-8' });
          },
        },
      ],
    };
    const result = await processFixture('puppet', options);
    expect(result).toMatchSnapshot();
  });
});
