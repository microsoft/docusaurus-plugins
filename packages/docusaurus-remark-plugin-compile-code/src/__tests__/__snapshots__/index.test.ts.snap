// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile-code plugin works on compile file 1`] = `
"Some snippets

\`\`\`lisp
foo
\`\`\`

    prefix

    foo
"
`;

exports[`compile-code plugin works on meta file 1`] = `
"Testing meta

\`\`\`fail skip foo
don't run this
\`\`\`

next, broken

\`\`\`fail ignore-errors foo
ignore errors
\`\`\`

    Error: fail

missing lang meta

\`\`\`fail
fail
\`\`\`

prefix

\`\`\`lisp prefix

\`\`\`

    prefix

\`\`\`lisp prefix
prefix
\`\`\`

    prefix

\`\`\`meta bar
baz
\`\`\`

    baz

\`\`\`meta
don't run this
\`\`\`
"
`;

exports[`compile-code plugin works on nodebin file 1`] = `
"Some snippets

\`\`\`ts
let x = 0
\`\`\`

next, broken

\`\`\`ts ignore-errors
sadfj asd fasdf
\`\`\`

    input.ts(1,1): error TS1434: Unexpected keyword or identifier.
    input.ts(1,7): error TS1434: Unexpected keyword or identifier.
"
`;

exports[`compile-code plugin works on nodes file 1`] = `
"Some snippets reflected in files

\`\`\`echonode
foo
\`\`\`

    extra --> foo

\`\`\`json
{
  "lang": "echonode",
  "meta": "",
  "cwd": ".docusaurus/docusaurus-remark-plugin-compile-code/cache/echonode/15c18188e8e42d94e4918be26812eeecda6b3bdc"
}
\`\`\`
"
`;

exports[`compile-code plugin works on puppet file 1`] = `
"\`\`\`puppet
hello world
\`\`\`

    browser says: hello world

\`\`\`puppet
hello world 2
\`\`\`

    browser says: hello world 2

\`\`\`puppet
hello
world
foo
bar
\`\`\`

    browser says: hello
    world
    foo
    bar
"
`;
